tests/test_addr.py

```diff
@@ -16,77 +17,69 @@ class AHTest(PyverbsAPITestCase):
-                ah_attr = AHAttr(gr=gr, is_global=1, port_num=port_num)
-                try:
-                    with AH(pd, attr=ah_attr):
-                        done += 1
-                except PyverbsRDMAError as ex:
-                    if ex.error_code == errno.EOPNOTSUPP:
-                        raise unittest.SkipTest('Create AH is not supported')
-                    raise ex
-        if done == 0:
-            raise unittest.SkipTest('No port is up, can\'t create AH')
-    # TODO: Test ibv_create_ah_from_wc once we have traffic
+            try:
+                AH(pd, attr=ah_attr)
+            except PyverbsRDMAError as ex:
+                if ex.error_code == errno.EOPNOTSUPP:
+                    raise unittest.SkipTest('Create AH is not supported')
+                raise ex
 
     def test_create_ah_roce(self):
         """
         Verify that AH can't be created without GRH in RoCE
         """
-        done = 0
-        for ctx, attr, attr_ex in self.devices:
+        for ctx, _, _ in self.devices:
+            self.verify_link_layer_ether(ctx)
+            self.verify_state(ctx)
             pd = PD(ctx)
-            for port_num in range(1, 1 + attr.phys_port_cnt):
-                port_attr = ctx.query_port(port_num)
-                if port_attr.state != e.IBV_PORT_ACTIVE and \
-                   port_attr.state != e.IBV_PORT_INIT:
-                    continue
-                if port_attr.link_layer != e.IBV_LINK_LAYER_ETHERNET:
-                    raise unittest.SkipTest('RoCE tests are only supported on Ethernet link layer')
-                ah_attr = AHAttr(is_global=0, port_num=port_num)
-                try:
-                    ah = AH(pd, attr=ah_attr)
-                except PyverbsRDMAError as ex:
-                    if ex.error_code == errno.EOPNOTSUPP:
-                        raise unittest.SkipTest('Create AH is not supported')
-                    assert 'Failed to create AH' in str(ex)
-                    done +=1
-                else:
-                    raise PyverbsError('Created a non-global AH on RoCE')
-        if done == 0:
-            raise unittest.SkipTest('No port is up, can\'t create AH')
+            ah_attr = AHAttr(is_global=0, port_num=self.ib_port)
+            try:
+                AH(pd, attr=ah_attr)
+            except PyverbsRDMAError as ex:
+                if ex.error_code == errno.EOPNOTSUPP:
+                    raise unittest.SkipTest('Create AH is not supported')
+                assert 'Failed to create AH' in str(ex)
+            else:
+                raise PyverbsError(f'Successfully created a non-global AH on RoCE port={self.ib_port}')
 
     def test_destroy_ah(self):
         """
         Test ibv_destroy_ah.
         """
-        done = 0
-        for ctx, attr, attr_ex in self.devices:
+        for ctx, _, _ in self.devices:
+            self.verify_state(ctx)
+            gr = u.get_global_route(ctx)
```

Comment:
I think you should pass the port_num here (looks like it was a bug also before your changes)

Reply:
Agree, Fixed.